### PR TITLE
Fix parsing floats panics and bugs

### LINKED ISSUE #1063: Panic parsing some floats
When running [this](https://github.com/tc39/test262/blob/main/test/language/types/number/8.5.1.js) test, we get the following panic:

```
thread 'main' panicked at 'Failed to parse float after checks: ParseFloatError { kind: Invalid }', boa/src/syntax/lexer/number.rs:384:50
``` 

This is the stack trace:
```
   0: rust_begin_unwind
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:495:5
   1: core::panicking::panic_fmt
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/option.rs:1268:5
   3: <boa::syntax::lexer::number::NumberLiteral as boa::syntax::lexer::Tokenizer<R>>::lex
   4: boa::syntax::lexer::Lexer<R>::next
   5: boa::syntax::parser::cursor::buffered_lexer::BufferedLexer<R>::fill
   6: boa::syntax::parser::cursor::buffered_lexer::BufferedLexer<R>::peek
   7: <boa::syntax::parser::expression::assignment::exponentiation::ExponentiationExpression as boa::syntax::parser::TokenParser<R>>::parse
   8: <boa::syntax::parser::expression::MultiplicativeExpression as boa::syntax::parser::TokenParser<R>>::parse
   9: <boa::syntax::parser::expression::AdditiveExpression as boa::syntax::parser::TokenParser<R>>::parse
  10: <boa::syntax::parser::expression::ShiftExpression as boa::syntax::parser::TokenParser<R>>::parse
  11: <boa::syntax::parser::expression::RelationalExpression as boa::syntax::parser::TokenParser<R>>::parse
  12: <boa::syntax::parser::expression::EqualityExpression as boa::syntax::parser::TokenParser<R>>::parse
  13: <boa::syntax::parser::expression::BitwiseANDExpression as boa::syntax::parser::TokenParser<R>>::parse
  14: <boa::syntax::parser::expression::BitwiseXORExpression as boa::syntax::parser::TokenParser<R>>::parse
  15: <boa::syntax::parser::expression::BitwiseORExpression as boa::syntax::parser::TokenParser<R>>::parse
  16: <boa::syntax::parser::expression::ShortCircuitExpression as boa::syntax::parser::TokenParser<R>>::parse
  17: <boa::syntax::parser::expression::assignment::conditional::ConditionalExpression as boa::syntax::parser::TokenParser<R>>::parse
  18: <boa::syntax::parser::expression::assignment::AssignmentExpression as boa::syntax::parser::TokenParser<R>>::parse
  19: <boa::syntax::parser::expression::Expression as boa::syntax::parser::TokenParser<R>>::parse
  20: <boa::syntax::parser::statement::Statement as boa::syntax::parser::TokenParser<R>>::parse
  21: <boa::syntax::parser::statement::StatementList as boa::syntax::parser::TokenParser<R>>::parse
  22: boa::context::Context::eval
  23: boa_tester::exec::<impl boa_tester::Test>::run_once
  24: boa_tester::exec::<impl boa_tester::Test>::run
  25: boa_tester::main
```

This happens [here](https://github.com/boa-dev/boa/blob/master/boa/src/syntax/lexer/number.rs#L384). In theory, at this point in the code, we have already checked that what we are receiving is a valid float literal, but it seems that some representation of it is passing the checks but still fails to parse.

The minimal code that seems to cause this issue is this one:
```javascript
4.9406564584124654417656879286822e-324;
```

This should in theory be parsed as `5e-324`, but instead, it panics. This might be because Rust doesn't parse floats the same way as JavaScript does, and we are using `f64::from_str`.

### PR DESCRIPTION
**Pull request by jevancc**
This Pull Request fixes/closes #1063 .

It changes the following:

- Add dependency `fast_float` to parse float string
- Fix parsing floats panic in number lexer
- Fix bugs in `parseFloat`, `parseInt`, and `Value::to_number`

### CODE REVIEW

**Comment by RageKnify** [comment]
### Test262 conformance changes:
  | Test result | master count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 78,497 | 78,497 | 0 |
  | Passed | 24,862 | 24,986 | **+124** |
  | Ignored | 15,587 | 15,587 | 0 |
  | Failed | 38,048 | 37,924 | **-124** |
  | Panics | 18 | 16 | **-2** |
  | Conformance | 31.67 | 31.83 | **+0.16%** |

Numbers look nice.

File: boa/src/builtins/string/mod.rs [hunk 1]
```diff
@@ -52,6 +52,27 @@ pub(crate) fn code_point_at(string: RcString, position: i32) -> Option<(u32, u8,
...
+    //
+    // Rust uses \p{White_Space} by default, which also includes:
+    // `\u{0085}' (next line)
+    // And does not include:
+    // '\u{FEFF}' (zero width non-breaking space)
+    // Explicit whitespace: https://tc39.es/ecma262/#sec-white-space
+    matches!(
+        c,
+        '\u{0009}' | '\u{000B}' | '\u{000C}' | '\u{0020}' | '\u{00A0}' | '\u{FEFF}' |
+    // Unicode Space_Seperator category
```
Comment (tofpie) [inline boa/src/builtins/string/mod.rs:68]:
```suggestion
    // Unicode Space_Separator category
```
Small typo

**Review by tofpie — APPROVED** [review]
Looks good to me! Thanks!

**Review by Lan2u — APPROVED** [review]
Looks really good

**Review by RageKnify — APPROVED** [review]

**Pull request merged**